src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -176,422 +227,317 @@ public void DisconnectFromWorkspace()
-
-                _notificationService.RegisterNotification(() =>
-                    {
-                        _workQueue.AssertIsForeground();
-                        ReportChangedSpan(changedSpan);
-                    },
-                    _listener.BeginAsyncOperation("EnqueueProcessSnapshotWorkerAsync.ReportChangedSpan"),
-                    _reportChangeCancellationSource.Token);
+                EnqueueWorkIfThisDocument(args.NewActiveContextDocumentId);
             }
 
-            private async Task<SnapshotSpan> GetChangedSpanAsync(
-                Document currentDocument, ITextSnapshot currentSnapshot, CancellationToken cancellationToken)
+            private void OnDocumentOpened(object? sender, DocumentEventArgs args)
             {
-                // We don't need to grab _lastProcessedDocument in a lock.  We don't care which version of the previous
-                // doc we grab, just that we grab some prior version.  This is only used to narrow down the changed range we 
-                // specify, so it's ok if it's slightly larger because we read in a change from a couple of edits ago.
-                var previousDocument = _lastProcessedDocument;
-                if (previousDocument != null)
-                {
-                    var service = TryGetClassificationService(currentSnapshot);
-                    if (service != null)
-                    {
-                        var changeRange = await service.ComputeSyntacticChangeRangeAsync(
-                            previousDocument, currentDocument, _diffTimeout, cancellationToken).ConfigureAwait(false);
-                        if (changeRange != null)
-                            return currentSnapshot.GetSpan(changeRange.Value.Span.Start, changeRange.Value.NewLength);
-                    }
-                }
-
-                // Couldn't compute a narrower range.  Just the mark the entire file as changed.
-                return currentSnapshot.GetFullSpan();
+                EnqueueWorkIfThisDocument(args.Document.Id);
             }
 
-            private void ReportChangedSpan(SnapshotSpan changeSpan)
+            private void OnWorkspaceChanged(object? sender, WorkspaceChangeEventArgs args)
             {
-                lock (_gate)
-                {
-                    var snapshot = _lastProcessedSnapshot;
-                    if (snapshot.Version.ReiteratedVersionNumber != changeSpan.Snapshot.Version.ReiteratedVersionNumber)
-                    {
-                        // wait for next call
-                        return;
-                    }
-                }
+                // We may be getting an event for a workspace we already disconnected from.  If so,
+                // ignore them.  We won't be able to find the Document corresponding to our text buffer,
+                // so we can't reasonably classify this anyways.
+                if (args.NewSolution.Workspace != _workspace)
+                    return;
```

Comment:
i'm not sure i understand this bit fully.  but i'm carryig it over from the old code.  seems terrifying.  but i think the presumption is that this is temporary and that WorkspaceRegistration far above will take care of disconnecting/reconnecting us to the new workspace.